### ISSUE TITLE
ComputationGraph topological sort is not guaranteed to be consistent across platforms

### ISSUE THREAD

**Issue by tail-island** [deeplearning4j/deeplearning4j#4057]
#### Issue Description

I'm writing an Android app using a neural-network imported from Keras 1.2.2.

Returned values from ComputationGraph#outputSingle() varies by CPU. On x86 Android emulator, output values are correct. But on ARM Android device, they are invalid.

Reproducing program is [here](https://github.com/tail-island/try-dl4j-android). Please run androidTest with Android emulator and ARM Android device. testWithComputationGraph() will fail on ARM only.

#### Version Information

* Deeplearning4j 9.0.1
* JavaCPP Presets for OpenBLAS 0.2.19-1.3
* Ubuntu 16.04
* ASUS ZenFone AR (Android 7.0)
* Android Studio 2.3.3 (Android Emulator 26.1.4, Android SDK Plargorm-Tools 26.0.0, Android SDK Tools 26.0.2)

#### Contributing

I think this issue's reason is below:

* Result of ComputationGraph#topologicalSortOrder() is different between x86 and ARM. The reason of difference is that HashMap iterating order is different between x86 and ARM.
* Neural-network will be initialized with invalid parameters. Because ComputationGraph#init() assigns parameters following topological sort order.
* If neural-network parameters are invalid, output values will be invalid.

Below Android app code outputs invalid values on ARM.

```
private void wrongOutputOnARM() throws IOException, ClassNotFoundException {
    // Adding reference. Because the model is imported from Keras and Reflections does not work on Android...
    NeuralNetConfiguration.reinitMapperWithSubtypes(Arrays.asList(new NamedType(TensorFlowCnnToFeedForwardPreProcessor.class)));

    // Loading model.
    ComputationGraph model = getModel(R.raw.model);

    // Prints output values. Correct on x86 but invalid on ARM...
    System.out.println(model.outputSingle(getInput(R.raw.input1), getInput(R.raw.input2), getInput(R.raw.input3), getInput(R.raw.input4)));
}

// Loading model.
private ComputationGraph getModel(int resourceId) throws IOException {
    try (InputStream stream = getResources().openRawResource(resourceId)) {
        return ModelSerializer.restoreComputationGraph(stream);
    }
}

// Loading input.
private INDArray getInput(int resourceId) throws IOException, ClassNotFoundException {
    try (InputStream stream = getResources().openRawResource(resourceId)) {
        return (INDArray)new ObjectInputStream(stream).readObject();
    }
}
```

Creating ComputationGraph wrapper that overrides topologicalSortOrder().

```
public class ComputationGraphWrapper extends ComputationGraph {
    public ComputationGraphWrapper(ComputationGraphConfiguration configuration) {
        super(configuration);
    }

    @Override
    public int[] topologicalSortOrder() {
        if (topologicalOrder != null)
            return topologicalOrder;

        (snip)

        // *** Changing the type of inputEdges from HashMap to LinkedHashMap for keeping order. ***
        Map<Integer, Set<Integer>> inputEdges = new LinkedHashMap<>(); //key: vertex. Values: vertices that the key vertex receives input from
        Map<Integer, Set<Integer>> outputEdges = new HashMap<>(); //key: vertex. Values: vertices that the key vertex outputs to

        (snip)

        return out;
    }
}
```

Using wrapper. Below code outputs correct values on ARM!

```
private void correctOutputOnX86AndARM() throws IOException, ClassNotFoundException {
    NeuralNetConfiguration.reinitMapperWithSubtypes(Arrays.asList(new NamedType(TensorFlowCnnToFeedForwardPreProcessor.class)));

    // Using ComputationGraphWrapper. Initialization code is copied from ModelSerializer.restoreComputationGraph().
    ComputationGraph model = new ComputationGraphWrapper(ComputationGraphConfiguration.fromJson(getContent(R.raw.configuration)));
    model.init(getCoefficients(R.raw.coefficients), false);

    // Prints output values. Correct on x86 and ARM!
    System.out.println(model.outputSingle(getInput(R.raw.input1), getInput(R.raw.input2), getInput(R.raw.input3), getInput(R.raw.input4)));
}

// Load contents.
private String getContent(int resourceId) throws IOException {
    try (BufferedReader reader = new BufferedReader(new InputStreamReader(getResources().openRawResource(resourceId)))) {
        StringBuilder sb = new StringBuilder();

        String line;
        while ((line = reader.readLine()) != null) {
            sb.append(line).append("\n");
        }

        return sb.toString();
    }
}

// Load parameters.
private INDArray getCoefficients(int resourceId) throws IOException, ClassNotFoundException {
    try (InputStream stream = getResources().openRawResource(resourceId)) {
        return Nd4j.read(new DataInputStream(stream));
    }
}
```

**Comment by saudet** [comment]
Thanks for looking into this!! Could you send a pull request with the suggested fix above?

**Comment by maxpumperla** [comment]
thanks @tail-island! Let us know if you need help with a PR, in case you have time for this.

@iRazvan would this help you for your problem?

**Comment by tail-island** [comment]
Thanks for checking my report! I have time and I had forked DL4J. But I'm not a good developer... Long time needed for creating build environment... Please wait a while.

One more. Maybe I cannot add new tests for this issue, because I don't know how to add Android test environment to DL4J. Is it okay?

**Comment by saudet** [comment]
@tail-island Yes that's alright, we don't have a test environment for Android yet...

**Comment by tail-island** [comment]
Thanks! I'm creating pull requests now. But I'm newbie; it's my first pull request. Please give me a time.

**Comment by maxpumperla** [comment]
@tail-island don't worry, take your time. we're here to help as well, just ping us if needed.

**Comment by AlexDBlack** [comment]
ComputationGraph is getting a substantial refactor here: https://github.com/deeplearning4j/deeplearning4j/pull/4050

I'll look at this at the same time as doing that. I already planned on adding the topological sort order to the JSON.

**Comment by tail-island** [comment]
Great! I'm looking forward to your refactored classes. But, may I fix this issue temporally? It's a chance for me contributing such a wonderful project.

**Comment by AlexDBlack** [comment]
Any changes there would have to be thoroughly tested to ensure backward compatibility.
You are welcome to submit a pull request, but I'll need to carefully review/test it before merging.

**Comment by tail-island** [comment]
Thanks a lot!

**Comment by raver119** [comment]
@AlexDBlack                        I guess this issue is fixed now?

**Comment by saudet** [comment]
No, I don't think so...

**Comment by AlexDBlack** [comment]
Right, it was fixed as part of an earlier large refactoring PR that wasn't completed.
I've been putting this off with the idea of getting back to that. But we'll probably split that up into a number of smaller/separate PRs instead.

tl;dr yes, it still needs to be fixed.

**Comment by AlexDBlack** [comment]
Fixed here: https://github.com/deeplearning4j/deeplearning4j/pull/5050
